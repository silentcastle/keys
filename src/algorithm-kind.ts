import { InvalidKeyKindError } from "./invalid-key-kind.error";

export enum AlgorithmKind {
  secp256k1 = "secp256k1",
  ed25519 = "ed25519",
  x25519 = "x25519",
}

export namespace AlgorithmKind {
  export function fromString(s: string) {
    const maybe = s as AlgorithmKind;
    switch (maybe) {
      case AlgorithmKind.secp256k1:
        return AlgorithmKind.secp256k1;
      case AlgorithmKind.ed25519:
        return AlgorithmKind.ed25519;
      case AlgorithmKind.x25519:
        return AlgorithmKind.x25519;
      default:
        throw new InvalidKeyKindError(maybe);
    }
  }
}
