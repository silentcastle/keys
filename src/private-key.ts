import { KeyKind } from "./key-kind";
import { IPublicKey } from "./public-key";

export interface IPrivateKey {
  readonly kind: KeyKind;
  publicKey(): Promise<IPublicKey>;
}

export interface ISigner {
  kind: KeyKind;
  sign(message: Uint8Array): Promise<Uint8Array>;
}

export interface ISignerIdentified extends ISigner {
  kid: string;
}
