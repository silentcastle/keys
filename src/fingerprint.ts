import * as multicodec from "multicodec";
import { IPublicKey } from "./public-key";
import { AlgorithmKind } from "./algorithm-kind";
import { InvalidKeyKindError } from "./invalid-key-kind.error";
import * as multibase from "multibase";

export enum KEY_PREFIX {
  secp256k1 = 0xe7,
  x25519 = 0xec,
  ed25519 = 0xed,
}

const decoder = new TextDecoder();

export function fingerprint(publicKey: IPublicKey): string {
  switch (publicKey.kind) {
    case AlgorithmKind.ed25519: {
      const bytes = multicodec.addPrefix(
        Uint8Array.from([KEY_PREFIX.ed25519]),
        publicKey.material
      );
      const encoded = multibase.encode("base58btc", bytes);
      return decoder.decode(encoded);
    }
    case AlgorithmKind.secp256k1: {
      const bytes = multicodec.addPrefix(
        Uint8Array.from([KEY_PREFIX.secp256k1]),
        publicKey.material
      );
      const encoded = multibase.encode("base58btc", bytes);
      return decoder.decode(encoded);
    }
    default:
      throw new InvalidKeyKindError(publicKey.kind);
  }
}
