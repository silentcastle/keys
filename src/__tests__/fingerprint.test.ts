import { PrivateKeyFactory } from "../private-key.factory";
import { AlgorithmKind } from "../algorithm-kind";
import { fingerprint } from "../fingerprint";

const factory = new PrivateKeyFactory();

test("secp256k1", async () => {
  const privateKey = factory.fromSeed(AlgorithmKind.secp256k1, "seed");
  const publicKey = await privateKey.publicKey();
  expect(fingerprint(publicKey)).toMatchSnapshot();
});

test("ed25519", async () => {
  const privateKey = factory.fromSeed(AlgorithmKind.ed25519, "seed");
  const publicKey = await privateKey.publicKey();
  expect(fingerprint(publicKey)).toMatchSnapshot();
});

test("x25519", async () => {
  const privateKey = factory.fromSeed(AlgorithmKind.x25519, "seed");
  const publicKey = await privateKey.publicKey();
  expect(fingerprint(publicKey)).toMatchSnapshot();
});
