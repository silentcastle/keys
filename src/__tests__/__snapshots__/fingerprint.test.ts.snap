// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ed25519 1`] = `"z6MkvhaQVvCfyWRovGMqgsKYDTxXsW2u7XQFyCTu8FLJ7JYV"`;

exports[`secp256k1 1`] = `"zQ3shawAiwRa3YbMitAcCyT9PhqPgy4Q4o1va8wzVyz9Lneh7"`;
